### ISSUE TITLE
Install Work Queue via Conda

### ISSUE THREAD

**Issue by PerilousApricot** [cooperative-computing-lab/cctools#1617]
I discussed briefly with @btovar the possibility of getting part of cctools put into pypi, so python-based tools which build on top of it can get what they need via pip's dependency resolution mechanisms. Currently lobster's install requires manually downloading/untarring the cctools binary and then updating PATH/LD_LIBRARY_PATH.

I'd like to work on the bits to get WQ into pypi automatically, which (if my research is right), would involve some small changes to the WQ repository, then when new releases are made, someone (?) would run a command to push the eggball into pypi. It may be able to push that into the existing procedure or alternately have something like Travis-CI monitor the repository for changes and automatically build/upload revisions.

**Comment by btovar** [comment]
There are parts of cctools used by lobster that do not relate to python at all. Is this a problem?
Moreover, I would not like to have two versions of cctools around, that is, the complete one, and a chopped one for pypi.

Could we achieve this without chopping or modifying cctools?  Say for example that the eggball never changes, and it simply contains a script that downloads the current version of cctools from our website, and installs it in the correct location? In this way, we can avoid the issue of the eggball being out of date.

**Comment by PerilousApricot** [comment]
> There are parts of cctools used by lobster that do not relate to python at all. Is this a problem?
It's not necessarily a problem, just that the python-ization (?) becomes more complicated the more that's added.

> Could we achieve this without chopping or modifying cctools?
It may be possible to do with little/no modification to existing cctools code, but depending on the guts of the distutils/setuptools, there would need to be new files added to the repository to instruct python how to install it

>  Say for example that the eggball never changes, and it simply contains a script that downloads the current version of cctools from our website, and installs it in the correct location?
If that were to happen, we'd want to push a new eggball for each released version, so it wouldn't be totally static.

Even without that, part of the reason (and niceness) of investigating this is to get the cctools shared libraries to properly link against non-system-pythons, which is a relatively common task in pythonland.

> In this way, we can avoid the issue of the eggball being out of date.
Regardless of the implementation details, someone (human, CI, release scripts) would need to hit a one-liner after each release to update the pypi registry. Once that's done, pip's regular dependency resolution stuff would grab the fresh release on a new install, or if someone requested their virtual env get upgraded.

**Comment by btovar** [comment]
CCTools is not properly linking to non-system pythons?

**Comment by PerilousApricot** [comment]
Man, GH really ate the formatting on my comment....

The bindings don't link because the python ABI changes between versions. For example, Python 3 can't use Python 2 bindings, etc...

**Comment by btovar** [comment]
python 3 is a whole-nother can of worms!

We even support python3 separately (e.g., --with-python3-path). I don't think lobster works with python3, does it?

I would be more worried if the python2 bindings do not work across installations. Is that the case?

**Comment by PerilousApricot** [comment]
That's also the case. CMS ships its own Python, which also struggles to
link properly.
-- 
It's dark in this basement.

**Comment by btovar** [comment]
Which kind of errors did you get? Could you send me an example of an environment and a command line?

**Comment by annawoodard** [comment]
I am not going to weigh in here on whether or not getting cctools into pypi is feasible, I leave that to you guys.

But in case you decide it isn't, as a consolation we can use the [scripts](http://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-scripts-keyword-argument) keyword from `setuptools` to automate installation of the cctools tarball. It doesn't give us the advantage of using pip's native dependency resolution and linking etc but it could at least get us to the hallowed `pip install lobster`.

**Comment by Nekel-Seyew** [comment]
@btovar Is this being actively looked into? I have a feeling that the idea won't really come to fruition, in which case, this can probably be closed.

**Comment by annawoodard** [comment]
@btovar I think we should consider re-opening this, considering Parsl/parsl#771

**Comment by dthain** [comment]
@eamonlm when you are ready, please share instructions here and make a PR with the Condor config files in `cctools/packaging/conda`

**Comment by dthain** [comment]
Condor -> Conda

**Comment by dthain** [comment]
@eamonlm please post the conda install instructions here, so interested parties can give it a test out.

**Comment by annawoodard** [comment]
I'm happy to test when this is ready. Very excited about this work!

**Comment by eamonlm** [comment]
Hello all, sorry for the delay. To install via conda, use the command "conda install -c cclnd cctools".

**Comment by annawoodard** [comment]
Conda installation went off without a hitch! Thanks, @eamonlm! Output below.

```
~ >conda create --name ccl Python=3.7
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /Users/awoodard/software/miniconda3/envs/ccl

  added / updated specs:
    - python=3.7


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2019.5.15  |                0         126 KB
    certifi-2019.6.16          |           py37_0         151 KB
    libcxx-4.0.1               |       hcfea43d_1         947 KB
    libcxxabi-4.0.1            |       hcfea43d_1         350 KB
    libedit-3.1.20181209       |       hb402a30_0         136 KB
    ncurses-6.1                |       h0a44026_1         732 KB
    openssl-1.1.1c             |       h1de35cc_1         2.2 MB
    pip-19.1.1                 |           py37_0         1.6 MB
    python-3.7.3               |       h359304d_0        18.0 MB
    setuptools-41.0.1          |           py37_0         506 KB
    sqlite-3.28.0              |       ha441bb4_0         1.2 MB
    wheel-0.33.4               |           py37_0          41 KB
    zlib-1.2.11                |       h1de35cc_3          90 KB
    ------------------------------------------------------------
                                           Total:        25.9 MB

The following NEW packages will be INSTALLED:

  ca-certificates    pkgs/main/osx-64::ca-certificates-2019.5.15-0
  certifi            pkgs/main/osx-64::certifi-2019.6.16-py37_0
  libcxx             pkgs/main/osx-64::libcxx-4.0.1-hcfea43d_1
  libcxxabi          pkgs/main/osx-64::libcxxabi-4.0.1-hcfea43d_1
  libedit            pkgs/main/osx-64::libedit-3.1.20181209-hb402a30_0
  libffi             pkgs/main/osx-64::libffi-3.2.1-h475c297_4
  ncurses            pkgs/main/osx-64::ncurses-6.1-h0a44026_1
  openssl            pkgs/main/osx-64::openssl-1.1.1c-h1de35cc_1
  pip                pkgs/main/osx-64::pip-19.1.1-py37_0
  python             pkgs/main/osx-64::python-3.7.3-h359304d_0
  readline           pkgs/main/osx-64::readline-7.0-h1de35cc_5
  setuptools         pkgs/main/osx-64::setuptools-41.0.1-py37_0
  sqlite             pkgs/main/osx-64::sqlite-3.28.0-ha441bb4_0
  tk                 pkgs/main/osx-64::tk-8.6.8-ha441bb4_0
  wheel              pkgs/main/osx-64::wheel-0.33.4-py37_0
  xz                 pkgs/main/osx-64::xz-5.2.4-h1de35cc_4
  zlib               pkgs/main/osx-64::zlib-1.2.11-h1de35cc_3


Proceed ([y]/n)? y


Downloading and Extracting Packages
libcxxabi-4.0.1      | 350 KB    | ############################################################################################ | 100%
setuptools-41.0.1    | 506 KB    | ############################################################################################ | 100%
wheel-0.33.4         | 41 KB     | ############################################################################################ | 100%
libcxx-4.0.1         | 947 KB    | ############################################################################################ | 100%
libedit-3.1.20181209 | 136 KB    | ############################################################################################ | 100%
sqlite-3.28.0        | 1.2 MB    | ############################################################################################ | 100%
zlib-1.2.11          | 90 KB     | ############################################################################################ | 100%
python-3.7.3         | 18.0 MB   | ############################################################################################ | 100%
ncurses-6.1          | 732 KB    | ############################################################################################ | 100%
certifi-2019.6.16    | 151 KB    | ############################################################################################ | 100%
pip-19.1.1           | 1.6 MB    | ############################################################################################ | 100%
ca-certificates-2019 | 126 KB    | ############################################################################################ | 100%
openssl-1.1.1c       | 2.2 MB    | ############################################################################################ | 100%
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate ccl
#
# To deactivate an active environment, use
#
#     $ conda deactivate

~ >conda activate ccl
(ccl) ~ >conda install -c cclnd cctools
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /Users/awoodard/software/miniconda3/envs/ccl

  added / updated specs:
    - cctools


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    cyrus-sasl-2.1.26          |       hb48c43a_4         238 KB
    krb5-1.16.1                |       hddcf347_7         1.0 MB
    libdb-6.1.26               |       h0a44026_0        17.4 MB
    libntlm-1.5                |       h1de35cc_0          30 KB
    mysql-5.7.24               |       h56848d4_0        17.4 MB
    ------------------------------------------------------------
                                           Total:        36.2 MB

The following NEW packages will be INSTALLED:

  cctools            cclnd/osx-64::cctools-7.0.15RC2-py37_0
  cyrus-sasl         pkgs/main/osx-64::cyrus-sasl-2.1.26-hb48c43a_4
  krb5               pkgs/main/osx-64::krb5-1.16.1-hddcf347_7
  libdb              pkgs/main/osx-64::libdb-6.1.26-h0a44026_0
  libntlm            pkgs/main/osx-64::libntlm-1.5-h1de35cc_0
  mysql              pkgs/main/osx-64::mysql-5.7.24-h56848d4_0


Proceed ([y]/n)? y


Downloading and Extracting Packages
krb5-1.16.1          | 1.0 MB    | ############################################################################################ | 100%
mysql-5.7.24         | 17.4 MB   | ############################################################################################ | 100%
cyrus-sasl-2.1.26    | 238 KB    | ############################################################################################ | 100%
libntlm-1.5          | 30 KB     | ############################################################################################ | 100%
libdb-6.1.26         | 17.4 MB   | ############################################################################################ | 100%
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
(ccl) ~ >which work_queue_worker
/Users/awoodard/software/miniconda3/envs/ccl/bin/work_queue_worker
(ccl) ~ >python
Python 3.7.3 (default, Mar 27 2019, 16:54:48)
[Clang 4.0.1 (tags/RELEASE_401/final)] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import work_queue
>>> work_queue.__file__
'/Users/awoodard/software/miniconda3/envs/ccl/lib/python3.7/site-packages/work_queue.py'
```

**Comment by eamonlm** [comment]
Thank you @annawoodard!

**Comment by dthain** [comment]
@leeping would you like to give this a try?

**Comment by btovar** [comment]
Fixed with #2176.  Any further issues should be filed as individual bugs.
